Adding organization section (#29796)
keycloak/keycloak: docs/documentation/server_admin/topics/organizations/authenticating-members.adoc

[id="authenticating-members_{context}"]

=  Authenticating Members
[role="_abstract"]

> Authenticating Members

Reviewer: ```suggestion
Authenticating members
```

> == Understanding the Identity-First Login

Reviewer: ```suggestion
== Understanding the identity-first login
```

[id="authenticating-members_{context}"]

=  Authenticating members
[role="_abstract"]

> One of the main changes introduced by organizations when enabled to a realm is on how users authenticate.
> These changes are mainly about being able to recognize if a user is authenticating in the context of an organization
> and changing the authentication flow accordingly on a per-organization basis.

Reviewer: ```suggestion
When you enable organizations for a realm, user authentication is changed. A user is authenticated in the context of an organization, which changes the authentication flow on a per-organization basis.
```

Author: The user is not always authenticating in the context of the organization as it depends on `recognizing` if the user actually maps to an organization.

That is what I tried to make more explicit there.

Author: Perfect :)

When a realm is created, the authentication flows are automatically updated to enable specific steps to authenticate and onboard organization members. The authentication flows updated are:

* `browser`

* `first broker login`

The main change to the `browser` flow is that it defaults to an identity-first login so that users are identified before prompting for their credentials.

> In regard to the first broker login flow, the main change there is to automatically add the user as an organization member once they authenticate through the identity provider associated with an organization and successfully complete flow.

Reviewer: ```suggestion
Concerning the first broker login flow, the main change is automatically adding the users as organization members once they authenticate through the identity provider associated with an organization and successfully complete the flow.
```

> The decision to whether or not an identity-first login should happen is based on the availability of any organization in a realm.
> If no organizations exist yet, the user will follow the usual steps to authenticate using both username and password,
> or any other step configured in the browser flow. Otherwise, the user will be asked first for a username or email to continue authenticating to a realm.

Reviewer: ```suggestion
The choice to use an identity-first login or not depends on the existence of an organization in a realm.
If no organizations exist yet, the user follows the usual steps to authenticate using the username and password,
or any other step configured in the browser flow. Otherwise, the user is asked first for a username or email to continue authenticating to a realm.
```

The identity-first login main goal is to identify the user:

* Is the user an existing or a new user?
* Is the user a member of any organization within a realm?
* If an organization member, is the user linked to any identity provider associated with the organization?

> Depending on the outcome when identifying the user, the authentication flow will change to either proceed with the authentication
> by asking for the user's credentials or eventually redirect the user automatically to authenticate within the organization security
> boundaries through an identity provider.

Reviewer: ```suggestion
Depending on the outcome when identifying the user, the authentication flow changes to either proceed with authentication
by asking for the user's credentials or eventually redirect the user automatically to authenticate within the organization security
boundaries through an identity provider.
```

== Understanding the identity-first login

In addition to identifying the user once the username is provided, the identity-first login is also responsible to:

* Match an email domain to an organization
* Decide if the authentication flow should continue or not if an account already exists for the username provided
* Decide how the user should be authenticated depending on how the domains and the identity providers are configured to an organization

> * Seamless authenticate users through an identity provider associated with an organization if the email domain matches the domain set to the identity provider

Reviewer: ```suggestion
* Seamlessly authenticate users through an identity provider associated with an organization if the email domain matches the domain set to the identity provider
```

> The identity-first login provides the very same capabilities when using the usual login page with both username and
> password fields. The users can still self-register by clicking on the register link or choose from any of the identity or social
> brokers available at the realm not linked to any organization.

Reviewer: ```suggestion
The identity-first login provides the very same capabilities that are used by the usual login page with the username and
password fields. Users can still self-register by clicking the register link or choose any identity or social
broker that is not linked to an organization in that realm.
```

> If a user that does not exist tries to authenticate using an email domain that matches an organization domain,

Reviewer: ```suggestion
In the case of a user that does not exist, if that user tries to authenticate using an email domain that matches an organization domain,
```

> the identity-first login page will be shown again and indicate that the username provided is not valid.

Reviewer: ```suggestion
the identity-first login page appears again and with a message that the username provided is not valid.
```

> At this point, there is no reason to ask the user for credentials in a second step.

Reviewer: ```suggestion
At this point, no need exists to ask the user for credentials.
```

> There are several ways to register the user so that he can authenticate to the realm and eventually join an organization.

Reviewer: ```suggestion
Several options exist to register the user allowing that user to authenticate to the realm and join an organization.
```

> If the realm has the self-registration setting enabled, the user can click on the `Register` link at the identity-first login page

Reviewer: ```suggestion
If the realm has the self-registration setting enabled, the user can click the *Register* link at the identity-first login page
```

Reviewer: We use "click" not "click on"
Also, we use *bold* for items in the user interface

and create an account at the realm.
After that, the administrator can send an invitation link to the user or manually add the user as a member of an organization.

> Look at the <<_managing_members_,Managing members>> for more details.

Reviewer: ```suggestion
For more details, see <<_managing_members_,Managing members>>.
```

Reviewer: Use this format,  "For more details, see.........."

> If the organization has an identity provider without a domain set, and they are marked as public, they can also click on

Reviewer: ```suggestion
If the organization has an identity provider without a domain set, and they are marked as public, the user can also click 
```

Reviewer: who or what is "they" in "they are marked as public"

Author: It should be `it is marked` as it refers to identity provider ...

Reviewer: Okay updated. Thanks.

the identity provider link at the identity-first login page to automatically create an account and join an organization

> once they authenticate through the identity provider.

Reviewer: ```suggestion
once that authenticates through the identity provider.
```

Author: Changed this sentence here https://github.com/keycloak/keycloak/pull/29796/commits/e1f6c3ec37454adbf9c543fcdb33cfe7c7a66ba2

> Look at the <<_managing_identity_provider_,Managing identity providers>> for more details.

Author: Changed here https://github.com/keycloak/keycloak/pull/29796/commits/e1f6c3ec37454adbf9c543fcdb33cfe7c7a66ba2

> Similar to the above, if the organization has an identity provider set with one of the organization domains,
> the user will be automatically redirected to the identity provider if their email matches a specific domain to authenticate
> and automatically create an account and join an organization once the flow is completed.

Reviewer: ```suggestion
In a similar situation to the previous section, the organization may have an identity provider set with one of the organization domains.  In this situation, the user is automatically redirected to the identity provider (if their email matches a specific domain to authenticate)
and automatically create an account and join an organization once the flow is completed.
```

Author: Changed here https://github.com/keycloak/keycloak/pull/29796/commits/5cb5d38e5fd64867e534ff4f531a108572ccdd5d.

== Configuring existing authentication flows

> As previously mentioned, for new realms authentication flows are automatically updated with the necessary steps

Reviewer: ```suggestion
As previously mentioned for new realms, authentication flows are automatically updated with the necessary steps
```

. Move the newly added `My Organization` sub-flow to execute right after the `Identity Provider Redirector` execution step.
The main point here is that the sub-flow should happen before any other sub-flow or execution step that authenticates the
user using whatever credentials you support in your realm. Once added, change the `Requirement` to `Alternative`.
. Click on `Add sub-flow` in the `My Organization` sub-flow and give it a name such as `My Organization - Conditional`. Once added, change the `Requirement` to `Conditional`.
. Click on `Add condition` in the `My Organization - Conditional` sub-flow and select `Condition - user configured`. Once added, change the `Requirement` to `Required`.
. Click on `Add step` in the `My Organization - Conditional` sub-flow and select the `Organization Identity-First Login
` execution step. Once added, change the `Requirement` to `Alternative`.
. Bind the authentication flow to the `Browser` binding type.

Once you enable the <<_enabling_organization_,Organizations>> setting to the realm and create
at least a single organization, you should be able to see the identity-first login page and start using organizations
in your realm.

> 

Reviewer: Use `backticks` only for what you type or enter
Use *bold* for whatever appears in the UI.

Author: Changed here https://github.com/keycloak/keycloak/pull/29796/commits/b40f57be332a46f8648dac83d3cdfc8926ef5e6f.
